File: src/infrastructure/tcp/services/flash-news-service.ts
import { inject, injectable } from "@needle-di/core";
import { PacketWriter } from "../../../core/tcp/utils/packet-builder-util.ts";
import { sendPacket } from "../../../core/tcp/utils/session-helpers-util.ts";
import { ActiveGameSessionsService } from "./active-game-sessions-service.ts";

const FLASH_NEWS_COMMAND = 0x4a50;
const PREFIX_PADDING_LENGTH = 7;

/** Known subcommand values embedded in bytes 3–4 of the 0x4A50 payload. */
export const FlashNewsSubcommand = {
  /** Standard server/ticker message — message text is displayed. */
  SERVER_MESSAGE: 0x4a68,
  /** Emergency maintenance notice — message text is ignored by the client;
   *  `maintenanceTime` encodes the start time via an unknown formula. */
  EMERGENCY_MAINTENANCE: 0x4a71,
} as const;

export type FlashNewsSubcommandValue = typeof FlashNewsSubcommand[keyof typeof FlashNewsSubcommand];

export interface FlashNewsAnnouncement {
  message?: string;
  unknown1?: number;
  unknown2?: number;
  /** Two-byte subcommand that controls client behaviour.
   *  High byte → payload byte 3, low byte → payload byte 4.
   *  @see FlashNewsSubcommand */
  subcommandHex?: number;
  unknown5?: number;
  /** Maintenance start time (subcommand 0x4A71 only).
   *  The exact encoding formula is not yet known. */
  maintenanceTime?: number;
}

export interface FlashNewsBroadcastResult {
  recipients: number;
  payloadLength: number;
}

@injectable()
export class FlashNewsService {
  constructor(
    private readonly activeGameSessionsService = inject(ActiveGameSessionsService),
  ) {}

  buildPayload({
    message = "",
    unknown1 = 0x00,
    unknown2 = 0x00,
    subcommandHex = FlashNewsSubcommand.SERVER_MESSAGE,
    unknown5 = 0x01,
    maintenanceTime = 0x00,
  }: FlashNewsAnnouncement): Uint8Array {
    const subcmdHigh = (subcommandHex >> 8) & 0xff;
    const subcmdLow = subcommandHex & 0xff;
    const writer = new PacketWriter();
    writer
      .writeUint8(unknown1)
      .writeUint8(unknown2)
      .writeUint8(subcmdHigh)
      .writeUint8(subcmdLow)
      .writeUint8(unknown5)
      .writeUint8(maintenanceTime)
      .writePadding(PREFIX_PADDING_LENGTH)
      .writeBytes(new TextEncoder().encode(message));
    return writer.build();
  }

  async broadcast(
    announcement: FlashNewsAnnouncement,
  ): Promise<FlashNewsBroadcastResult> {
    const payload = this.buildPayload(announcement);
    const sessions = this.activeGameSessionsService.list();

    await Promise.all(
      sessions.map((session) => sendPacket(session, FLASH_NEWS_COMMAND, payload)),
    );

    return {
      recipients: sessions.length,
      payloadLength: payload.length,
    };
  }
}

File: src/Shared/Domain/News/FlashNewsService.cs
using Mgo2Server.Shared.Tcp;
using Mgo2Server.Shared.Utils;

namespace Mgo2Server.Shared.Domain.News;

/// <summary>Subcommand values carried in bytes three and four of the ticker payload.</summary>
public static class FlashNewsSubcommand
{
    /// <summary>Server ticker message; the client displays the message text.</summary>
    public const ushort ServerMessage = 0x4a68;

    /// <summary>Emergency maintenance notice; the client ignores the message text.</summary>
    public const ushort EmergencyMaintenance = 0x4a71;
}

/// <summary>One ticker announcement.</summary>
/// <param name="Message">Text shown in the client's ticker.</param>
/// <param name="Unknown1">First byte of the leading field, whose purpose is unknown.</param>
/// <param name="Unknown2">Second byte of the leading field, whose purpose is unknown.</param>
/// <param name="Subcommand">Subcommand that selects the client behaviour.</param>
/// <param name="Unknown5">Fifth byte of the payload, whose purpose is unknown.</param>
/// <param name="MaintenanceTime">Maintenance start time, used by the emergency subcommand.</param>
public sealed record FlashNewsAnnouncement(
    string Message = "",
    byte Unknown1 = 0x00,
    byte Unknown2 = 0x00,
    ushort Subcommand = FlashNewsSubcommand.ServerMessage,
    byte Unknown5 = 0x01,
    byte MaintenanceTime = 0x00);

/// <summary>Outcome of a ticker broadcast.</summary>
/// <param name="Recipients">Number of clients the announcement was written to.</param>
/// <param name="PayloadLength">Length of the payload that was written.</param>
public sealed record FlashNewsBroadcastResult(int Recipients, int PayloadLength);

/// <summary>
/// Builds the ticker packet and writes it to the gameplay lobby clients the
/// server currently serves.
/// </summary>
/// <param name="activeGameSessions">Sessions of the running lobby instances.</param>
/// <param name="sessionHelper">Helper used to write the packets.</param>
public sealed class FlashNewsService(
    ActiveGameSessionsService activeGameSessions,
    SessionHelper sessionHelper)
{
    /// <summary>Command the ticker is written with.</summary>
    private const ushort FlashNewsCommand = 0x4a50;

    /// <summary>Zero-filled padding that precedes the message text.</summary>
    private const int PrefixPaddingLength = 7;

    /// <summary>Builds the payload of one announcement.</summary>
    /// <param name="announcement">Announcement to encode.</param>
    public static byte[] BuildPayload(FlashNewsAnnouncement announcement)
    {
        var writer = new PacketWriter();
        writer
            .WriteUInt8(announcement.Unknown1)
            .WriteUInt8(announcement.Unknown2)
            .WriteUInt8(announcement.Subcommand >> 8)
            .WriteUInt8(announcement.Subcommand & 0xff)
            .WriteUInt8(announcement.Unknown5)
            .WriteUInt8(announcement.MaintenanceTime)
            .WritePadding(PrefixPaddingLength)
            .WriteBytes(System.Text.Encoding.UTF8.GetBytes(announcement.Message));

        return writer.Build();
    }

    /// <summary>Broadcasts an announcement to every client of this process's lobbies.</summary>
    /// <param name="announcement">Announcement to broadcast.</param>
    /// <param name="cancellationToken">Token that cancels the operation.</param>
    public async Task<FlashNewsBroadcastResult> BroadcastAsync(
        FlashNewsAnnouncement announcement,
        CancellationToken cancellationToken = default)
    {
        var payload = BuildPayload(announcement);
        var sessions = activeGameSessions.List();

        foreach (var session in sessions)
        {
            await sessionHelper.SendPacketAsync(session, FlashNewsCommand, payload, cancellationToken);
        }

        return new FlashNewsBroadcastResult(sessions.Count, payload.Length);
    }
}
